lib/src/metta/runner/builtin_mods/fileio.rs

```diff
@@ -0,0 +1,168 @@
+use std::cell::RefCell;
+use std::fmt::{Display, Formatter};
+use crate::{sym, Atom, ExecError, Grounded};
+use crate::metta::{ARROW_SYMBOL, UNIT_ATOM};
+use crate::metta::runner::stdlib::unit_result;
+use crate::metta::runner::str::{Str, ATOM_TYPE_STRING};
+use std::fs;
+use std::fs::OpenOptions;
+use std::io::{Read, Seek, SeekFrom, Write};
+use std::rc::Rc;
+use crate::space::grounding::GroundingSpace;
+use crate::metta::text::SExprParser;
+use crate::metta::runner::{ModuleLoader, RunContext, DynSpace, Metta, MettaMod};
+use crate::atom::gnd::*;
+
+pub static FILEIO_METTA: &'static str = include_str!("fileio.metta");
+pub const ATOM_TYPE_FILE_HANDLE: Atom = sym!("FileHandle");
+
+#[derive(Clone, Debug)]
+pub struct FileHandle(Rc<RefCell<fs::File>>);
+
+impl FileHandle {
+    fn open(path: &str, options: &str) -> Result<Self, ExecError> {
+
+        let mut opened_file = OpenOptions::new();
+        let opened_file= opened_file
+            .read(options.contains("r"))
+            .write(options.contains("w"))
+            .create(options.contains("c"))
+            .append(options.contains("a"))
+            .truncate(!options.contains("a"));
+
+        match opened_file.open(path)
+        {
+            Ok(file) => Ok(Self(Rc::new(RefCell::new(file)))),
+            Err(_) => Err(ExecError::from(format!("Failed to open file with provided path={} and options={}", path, options)))
+        }
+    }
+
+    fn read_as_str(&self) -> Result<String, ExecError>
+    {
+        self.0.borrow_mut().seek(SeekFrom::Start(0)).unwrap();
+        let mut contents = String::new();
+        match self.0.borrow_mut().read_to_string(&mut contents) {
+            Ok(_) => Ok(contents),
+            Err(_) => Err(ExecError::from("Failed to read file contents"))
+        }
+    }
+
+    fn write(&self, content: &str) -> Result<(), ExecError>
+    {
+        match self.0.borrow_mut().write(&content.as_bytes()) {
+            Ok(_) => Ok(()),
+            Err(_) => Err(ExecError::from("Failed to read file contents"))
```

Comment:
It is a write function:
```suggestion
            Err(_) => Err(ExecError::from("Failed to write file contents"))
```
Also it is worth to add the reason of the error into the error message (for both read and write). Because user should understand what caused the issue to prevent it next time.

Reply:
> Also it is worth to add the reason of the error into the error message (for both read and write). Because user should understand what caused the issue to prevent it next time.

You mean something like that?
`Err(message) => Err(ExecError::from(format!("Failed to write file contents: {}", message)))`

Reply:
Yeah